django_ses/utils.py

```diff
@@ -308,3 +309,32 @@ def confirm_sns_subscription(notification):
             },
             exc_info=True,
         )
+
+
+def get_permanent_bounced_emails_from_bounce_obj(bounce_obj: dict) -> list:
+    """Extracts permanent bounced email addresses only as a list of strings.
+    https://docs.aws.amazon.com/ses/latest/DeveloperGuide/notification-contents.html#bounce-object
+    """
+    bounced_recipients = bounce_obj.get("bouncedRecipients", list())
+    return [br.get("emailAddress") for br in bounced_recipients if br.get("status").startswith("5")]
+
+
+def get_emails_from_complaint_obj(complaint_obj: dict) -> list:
+    """Extracts complaint email addresses from complaint_obj
+    https://docs.aws.amazon.com/ses/latest/DeveloperGuide/notification-contents.html#complaint-object
+    """
+    return [cr.get("emailAddress") for cr in complaint_obj.get("complainedRecipients", list())]
+
+
+def filter_blacklisted_recipients(addresses):
+    """Remove blacklisted emails from addresses"""
+    if isinstance(addresses, str):
+        addresses = [addresses]
+
+    filtered_addresses = []
+    for recipient in addresses:
+        addr = parseaddr(recipient)[1].lower()
+        if models.BlacklistedEmail.objects.filter(email=addr).exists():
+            continue
```

Comment:
Would prefer to do a single lookup instead of one for each recipient. Should be able to do it like this:

```python
emails = [parseaddr(recipient)[1].lower() for recipient in addresses]
excluded_emails = set(models.BlacklistedEmail.objects.filter(email__in=emails).values_list("email"))
return [email for email in emails if email not in excluded_emails]
```

Remember to account for the lowercase setting I mention in my other comment.

Reply:
Indeed, this will be better for the performance 🙏 will change! (awaiting the lowercase thingy until you reply in the other thread)

Reply:
I replied and am OK to skip the lowercase flag, keeping the lowercase logic you have.